src/libsyntax/parse/lexer/mod.rs

```diff
@@ -1701,71 +1395,62 @@ impl<'a> StringReader<'a> {
     }
 
     fn scan_byte(&mut self) -> token::Lit {
+        let start_with_quote = self.pos;
         self.bump();
         let start = self.pos;
 
         // the eof will be picked up by the final `'` check below
         let c2 = self.ch.unwrap_or('\x00');
         self.bump();
 
-        let valid = self.scan_char_or_byte(start,
-                                           c2,
-                                           // ascii_only =
-                                           true,
-                                           '\'');
-        if !self.ch_is('\'') {
-            // Byte offsetting here is okay because the
-            // character before position `start` are an
-            // ascii single quote and ascii 'b'.
-            let pos = self.pos;
-            self.fatal_span_verbose(start - BytePos(2),
-                                    pos,
-                                    "unterminated byte constant".to_string()).raise();
+        if c2 == '\\' && (self.nextch_is('\'') || self.nextch_is('\\')) {
+            self.bump();
+        }
+        if c2 == '\'' && !self.ch_is('\'') {
+            let id = self.name_from_to(start, start);
+            return token::Byte(id);
         }
 
-        let id = if valid {
-            self.name_from(start)
-        } else {
-            Symbol::intern("?")
-        };
-        self.bump(); // advance ch past token
+        loop {
+            if self.ch_is('\'') {
+                break;
+            }
+            if self.ch_is('\\') && (self.nextch_is('\'') || self.nextch_is('\\')) {
+                self.bump();
+                self.bump();
+            } else {
+                // Only attempt to infer single line string literals. If we encounter
+                // a slash, bail out in order to avoid nonsensical suggestion when
+                // involving comments.
+                if self.is_eof() || self.ch_is('/') || self.ch_is('\n') {
+                    self.err_span(self.mk_sp(start_with_quote, self.pos),
+                                  "unterminated byte constant");
+                    return token::Err(self.name_from(start));
+                }
+                self.bump();
+            }
+        }
```

Comment:
Looks like the loop for character literals above is very similar, perhaps they can be unified.